cleo_plugins/Audio/Audio.cpp

```diff
@@ -0,0 +1,320 @@
+    static OpcodeResult __stdcall opcode_0AC2(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+        CVector pos;
+        pos.x = OPCODE_READ_PARAM_FLOAT();
+        pos.y = OPCODE_READ_PARAM_FLOAT();
+        pos.z = OPCODE_READ_PARAM_FLOAT();
+
+        stream->Set3dPosition(pos);
+        return OR_CONTINUE;
+    }
+
+    //0AC3=2,link_3d_audiostream %1d% to_object %2d%
+    static OpcodeResult __stdcall opcode_0AC3(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+        auto handle = OPCODE_READ_PARAM_OBJECT_HANDLE();
+
+        auto object = CPools::GetObject(handle);
+        stream->Link(object);
+
+        return OR_CONTINUE;
+    }
+
+    //0AC4=2,link_3d_audiostream %1d% to_actor %2d%
+    static OpcodeResult __stdcall opcode_0AC4(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+        auto handle = OPCODE_READ_PARAM_PED_HANDLE();
+
+        auto ped = CPools::GetPed(handle);
+        stream->Link(ped);
+
+        return OR_CONTINUE;
+    }
+
+    //0AC5=2,link_3d_audiostream %1d% to_vehicle %2d%
+    static OpcodeResult __stdcall opcode_0AC5(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+        auto handle = OPCODE_READ_PARAM_VEHICLE_HANDLE();
+
+        auto vehicle = CPools::GetVehicle(handle);
+        stream->Link(vehicle);
+
+        return OR_CONTINUE;
+    }
+
+    //2500=1,  is_audio_stream_playing %1d%
+    static OpcodeResult __stdcall opcode_2500(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+
+        auto state = stream->GetState();
+
+        OPCODE_CONDITION_RESULT(state == 1);
+        return OR_CONTINUE;
+    }
+
+    //2501=2,%2d% = get_audiostream_length_fine %1d%
+    static OpcodeResult __stdcall opcode_2501(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+
+        auto length = stream->GetLength();
+        length /= max(stream->GetSpeed(), 0.000001f); // speed corrected
+
+        OPCODE_WRITE_PARAM_FLOAT(length);
+        return OR_CONTINUE;
+    }
+
+    //2502=2,get_audio_stream_speed %1d% store_to %2d%
+    static OpcodeResult __stdcall opcode_2502(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+
+        auto speed = stream->GetSpeed();
+
+        OPCODE_WRITE_PARAM_FLOAT(speed);
+        return OR_CONTINUE;
+    }
+
+    //2503=2,set_audio_stream_speed %1d% speed %2d%
+    static OpcodeResult __stdcall opcode_2503(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+        auto speed = OPCODE_READ_PARAM_FLOAT();
+
+        stream->SetSpeed(speed);
```

Comment:
probably should disallow speed of 0. one should pause stream instead

Reply:
Intended. There is opcode allowing to perform smooth speed transition. So transitioning to 0.0 is useful to stop the audio with "tape scratch" effect.